File: 1. Two Sum.py
# 1. Two Sum
# Time:  O(N)
# Space: O(N)

# use a map to track number index mappings
class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        nums_index_map = {}
        for i, num in enumerate(nums):
            to_find = target - num
            if to_find in nums_index_map:
                return [nums_index_map[to_find], i]
            else:
                nums_index_map[num] = i

File: 1. Two Sum.java
// 1. Two Sum
// Time:  O(n)
// Space: O(n)

// HashMap to keep track of current to find number
public class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> m = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int toFind = target - nums[i];
            if (m.containsKey(toFind)) {
                return new int[] { m.get(toFind), i };
            }
            m.put(nums[i], i);
        }
        throw new IllegalArgumentException("No two sum solution");
    }
}
